rust/tw_evm/src/abi/encode.rs

```diff
@@ -1083,4 +1143,124 @@ mod tests {
         .unwrap();
         assert_eq!(encoded, expected);
     }
-}
+
+    #[test]
+    fn encode_packed_address() {
+        let address = Token::Address("0x1111111111111111111111111111111111111111".into());
+        let encoded = encode_packed_tokens(&[address]);
+        let expected = "1111111111111111111111111111111111111111".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+    }
+
+    #[test]
+    fn encode_packed_bytes() {
+        let bytes = Token::Bytes(vec![0x12, 0x34]);
+        let encoded = encode_packed_tokens(&[bytes]);
+        let expected = "1234".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+    }
+
+    #[test]
+    fn encode_packed_string() {
+        let s = Token::String("Hello, world!".to_owned());
+        let encoded = encode_packed_tokens(&[s]);
+        let expected = "48656c6c6f2c20776f726c6421".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+    }
+
+    #[test]
+    fn encode_packed_int() {
+        let int = Token::Int {
+            int: I256::from(-1),
+            bits: UintBits::new(256).unwrap(),
+        };
+        let encoded = encode_packed_tokens(&[int]);
+        let expected = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+    }
+
+    #[test]
+    fn encode_packed_bool() {
+        let b = Token::Bool(true);
+        let encoded = encode_packed_tokens(&[b]);
+        let expected = "01".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+
+        let b = Token::Bool(false);
+        let encoded = encode_packed_tokens(&[b]);
+        let expected = "00".decode_hex().unwrap();
+        assert_eq!(encoded, expected);
+    }
+
+    #[test]
+    fn encode_packed_uint() {
+        let uint = Token::Uint {
+            uint: U256::from(3_u64),
+            bits: UintBits::new(256).unwrap(),
+        };
+        let encoded = encode_packed_tokens(&[uint]);
+        let expected = "03".decode_hex().unwrap();
```

Comment:
I've updated everything else, except this. I would need further light on this one.
I've got between:

```rust
Token::Int { int, bits } => {
                let buf = int.to_big_endian().as_slice().to_vec();
                let bits_value = bits.get();
                let padding_len = (bits_value / 8) - buf.len();
                let padding = vec![0; padding_len];
                encoded.extend_from_slice(&padding);
                encoded.extend_from_slice(&buf);
            },
            Token::Uint { uint, bits } => {
                let buf = uint.to_big_endian().as_slice().to_vec();
                let bits_value = bits.get();
                let padding_len = (bits_value / 8) - buf.len();
                let padding = vec![0; padding_len];
                encoded.extend_from_slice(&padding);
                encoded.extend_from_slice(&buf);
            },
```

and:

```rust
Token::Uint { uint, bits } => {
    let full_length = bits.get() / 8; // Calculate the full byte width
    let mut buf = vec![0; full_length]; // Create a buffer with zeros
    uint.to_big_endian(&mut buf[(full_length - buf.len())..]); // Fill the buffer from the right
    encoded.extend_from_slice(&buf); // Add to the encoded vector
},
```

What are your thoughts?

Reply:
I believe the first way wouldn't work because `int.to_big_endian()` returns 32 bytes always. But you can use `int.to_big_endian_compact()` as it removes leading zeros. Furthermore, there is `int.to_big_endian_compact_min_len()` function that takes the minimum required length, and the result will be padded with zero bytes.